dfiq.py

```diff
@@ -0,0 +1,523 @@
+            self.is_internal = True
+
+
+class Question(Component):
+    """A Question in DFIQ.
+
+    Questions are the fundamental "building blocks" of DFIQ. All other DFIQ
+    components are relative to Questions: Approaches describe how to answer the
+    Questions, and Scenarios and Facets organize the Questions logically.
+
+    Attributes:
+        type (str): The type of component, which is always "question".
+    """
+
+    def __init__(self, dfiq_id, name, description, tags, parent_ids):
+        super().__init__(dfiq_id, name, description, tags, parent_ids)
+        self.type = "question"
+
+    @property
+    def approaches(self):
+        return self.child_ids
+
+
+class Facet(Component):
+    """A Facet in DFIQ.
+
+    Facets are used for intermediate-level grouping in DFIQ. A particular Facet can
+    be part of multiple different Scenarios and will contain multiple Questions. A
+    Facet breaks the larger Scenario into smaller logical pieces, but a Facet is
+    still too broad to answer directly; it must also be broken down (into Questions).
+
+    Attributes:
+        type (str): The type of component, which is always "facet".
+    """
+
+    def __init__(self, dfiq_id, name, description, tags, parent_ids):
+        super().__init__(dfiq_id, name, description, tags, parent_ids)
+        self.type = "facet"
+
+    @property
+    def questions(self):
+        return self.child_ids
+
+
+class Scenario(Component):
+    """A Scenario in DFIQ.
+
+    A Scenario is the highest-level grouping in DFIQ. A Scenario is made of one or
+    more Facets (different "sides" of an investigation), which in turn are made up of
+    investigative Questions.
+
+    Attributes:
+        type (str): The type of component, which is always "scenario".
+    """
+
+    def __init__(self, dfiq_id, name, description, tags):
+        super().__init__(dfiq_id, name, description, tags)
+        self.type = "scenario"
+
+    @property
+    def facets(self):
+        return self.child_ids
+
+
+class DFIQ:
+    """A DFIQ knowledge base.
+
+    An instance of a DFIQ knowledge base. Upon initialization, it reads DFIQ YAML files from
+    yaml_data_path, converts them to Python objects, and builds a graph of the components'
+    relationships.
+
+    Attributes:
+        yaml_data_path (str): The path to the directory containing the DFIQ YAML files.
+        markdown_output_path (str, optional): The path to the directory where the
+            generated Markdown files should be saved.
+        plural_map (dict): A dictionary mapping from DFIQ component types to their
+            plural forms.
+        components (dict): A dictionary mapping from DFIQ component IDs to their
+            corresponding components.
+        graph (nx.DiGraph, optional): A directed graph representing the relationships
+            between DFIQ components.
+        jinja_env (jinja2.Environment): A Jinja2 environment used to generate Markdown
+            files.
+
+    Methods:
```

Comment:
Why not use doc strings on the methods below instead?

Reply:
Moved